Add manual retriggering of bundle rebuild

apaplaus opened the pull request:
Now when there is need in retriggering bundle build due to fail or
something else, there is functionality for it.

Note: it's implemented only to retrigger automatic rebuilds, it's not
for manual rebuild event triggered by BOTAS.

JIRA: CWFHEALTH-603

Signed-off-by: Andrei Paplauski <apaplaus@redhat.com>

qixiang commented on freshmaker/parsers/internal/manual_rebuild.py:
```diff
@@ -51,10 +52,16 @@ def parse_post_data(self, data):
+            # BotasErrataShippedEvent event
+            if advisory.state == "SHIPPED_LIVE" and \
+                    advisory.reporter.startswith('botas'):
+                event = BotasErrataShippedEvent(msg_id, advisory)
```
I believe we need different event to distinguish the original `BotasErrataShippedEvent` and the re-triggered one, so reusing `BotasErrataShippedEvent` is not what we want.

qixiang reviewed (commented):
The first thing need to be addressed is Freshmaker should skip the succeed builds in original event.

apaplaus commented:
> The first thing need to be addressed is Freshmaker should skip the succeed builds in original event.

Yes, probably it would be better to skip successful builds from the previous event. But I am trying to find parts of the code where we are doing this for the manual rebuilds for advisories. And I can't find any neither for ManualRebuildWithAdvisoryEvent nor for FreshmakerAsyncManualBuildEvent. Can you please help me with finding it?

qixiang commented:
> > The first thing need to be addressed is Freshmaker should skip the succeed builds in original event.
> 
> Yes, probably it would be better to skip successful builds from the previous event. But I am trying to find parts of the code where we are doing this for the manual rebuilds for advisories. And I can't find any neither for ManualRebuildWithAdvisoryEvent nor for FreshmakerAsyncManualBuildEvent. Can you please help me with finding it?

iirc, it's handled by:
https://github.com/redhat-exd-rebuilds/freshmaker/blob/master/freshmaker/handlers/koji/rebuild_images_on_rpm_advisory_change.py#L231-L247
https://github.com/redhat-exd-rebuilds/freshmaker/blob/28723e72c59c026996c03b4019c9489f3d8d90a2/freshmaker/models.py#L497-L510

apaplaus commented:
@qixiang well I've added the new event, and added filtering out already done builds from the dependent event. I think the new event can be handled exactly the same way as the automatic rebuild, since we are processing the advisory, and the difference will be in ignoring builds from the dependent event. But I am not sure =) So can you take a look again pls?

qixiang commented on freshmaker/parsers/internal/manual_rebuild.py:
```diff
@@ -51,10 +52,17 @@ def parse_post_data(self, data):
+                    freshmaker_event_id=data.get('freshmaker_event_id'), manual=True,
+                    dry_run=dry_run, requester=data.get('requester', None))
         else:
             event = ManualBundleRebuild(msg_id, data.get('container_images', []),
```
`ManualBundleRebuild` now doesn't support advisory id, this should be adjusted, release-driver will also need the support of advisory id in `ManaualBundleRebuild` event.

apaplaus replied:
Yes, and it was the part of the implementation, I was pretty sure during the implementation of this event, that there will be no advisory submitted/passed from BOTAS, so BOTAS will submit ONLY bundles + container_images, so I should work with only submitted information.

apaplaus commented on freshmaker/events.py:
```diff
@@ -472,3 +473,11 @@ def __init__(self, msg_id, container_images, bundle_images, metadata=None,
         return super().is_allowed(handler, ArtifactType.IMAGE, **kwargs)
+
+
+class RetriggeredBundleRebuild(ErrataBaseEvent):
```
Well when Matt was describing this task he said that this 'retriggering' functionality should work in almost the same way as it works now for manually retriggering failed event of rhsa advisory rebuild. And it works in that way that you can submit post request with the `freshmaker_event_id` of failed event. FM will search for it -> create new db_event with the dependency set to the failed event  + set errata_id from the failed event so we have an advisory -> rebuild images + while rebuilding since it has dependent event it will not rebuild successfully built images from it. Right? So I thought if we want the same functionality(just submit post request with eighter errata_id OR freshmaker_event_id) I should do it like I did. So now if you submit POST request with the `errata_id` or with the `freshmaker_event_id`, it will trigger the rebuild like it handles advisory, and the difference will be that if `freshmaker_event_id` is submitted it will add dependent event and will ignore done builds. Maybe I missed something in implementation, but that's the other question. And about `ManualBundleRebuild`, it's an event triggered exclusively by BOTAS sending the request(maybe it's not the best name for the event then but still), and BOTAS always submits `bundle_images` as well as `container_images` and there is no advisory at all in the processing process.

qixiang replied:
> Well when Matt was describing this task he said that this 'retriggering' functionality should work in almost the same way as it works now for manually retriggering failed event of rhsa advisory rebuild. And it works in that way that you can submit post request with the `freshmaker_event_id` of failed event. FM will search for it -> create new db_event with the dependency set to the failed event + set errata_id from the failed event so we have an advisory -> rebuild images + while rebuilding since it has dependent event it will not rebuild successfully built images from it. Right? So I thought if we want the same functionality(just submit post request with eighter errata_id OR freshmaker_event_id) I should do it like I did. So now if you submit POST request with the `errata_id` or with the `freshmaker_event_id`, it will trigger the rebuild like it handles advisory, and the difference will be that if `freshmaker_event_id` is submitted it will add dependent event and will ignore done builds.

All right.

> Maybe I missed something in implementation, but that's the other question. And about `ManualBundleRebuild`, it's an event triggered exclusively by BOTAS sending the request(maybe it's not the best name for the event then but still), and BOTAS always submits `bundle_images` as well as `container_images` and there is no advisory at all in the processing process.

How about just updating `ManualBundleRebuild` to be triggered by either `errata_id` or `freshmaker_event_id` or `bundle_images`?

apaplaus replied:
So you want to have one event to represent two different situation. One- is when botas submits `bundle images` and it handled in one way, and Second - when event submitted manually with the `errata_id/freshmaker_event_id` and then it handled in another way? Seem a little bit like overcomplicating, you will not know what was the trigger of the event just by it type, you will need to check other stuff. For me it's much easier just to maybe rename current event to kind of 'BotasTriggeredRebuild' or something else and use 'ManualBundleRebuild' for the purpose of manually triggered reuilds(with `errata_id/freshmaker_event_id`) What do you think?

qixiang replied:
It's release-driver (not botas) triggers the manual rebuild.

They're both manual rebuilds, though one is triggered by a service (release-driver), another one is triggered by human beings, this is same as the existing manual rebuild of rpm advisories, where release-driver requests manual rebuilds with the same event used by us to trigger by manual, though release-driver's request usually contains some additional metadata.

Actually, release-driver can always provide the `errata_id` in request, so you just need to check the advisory (from `errata_id` or `freshmaker_event_id`) to determine whether this should be handled by the botas handler, and in the handler, you need to handle a special request that has `bundle_images` included, what do you think?

apaplaus replied:
But I don't need `errata_id` at all when RD triggers rebuild, and now the flow is very different when an event has bundle_images(+container_images) and when an event has errata_id(+ freshamker_event_id)

qixiang replied:
I'm thinking of dropping the `bundle_images` in manual rebuild request to simplify the workflow, seems it's possible to achieve the same goal of `bundle_images` + `container_images` with `errata_id` + `container_images`.

With `bundle_images`, freshmaker can get the pulspec overrides from its db, and with `errata_id`, freshmaker can get a mapping of old digests to new digests, we can repeat the construction of pullspec overrides, what do you think?

apaplaus replied:
I think that I don't want to do the additional work of reimplementing manual rebuilds triggered by RD in the scope of the current ticket, because it will prolong this ticket more(it's already pretty long-lasting as for me). If we want to change it, I would prefer another ticket so the work is tracked.

qixiang replied:
> I think that I don't want to do the additional work of reimplementing manual rebuilds triggered by RD in the scope of the current ticket, because it will prolong this ticket more(it's already pretty long-lasting as for me). If we want to change it, I would prefer another ticket so the work is tracked.

ok, then how about at least reusing the existing `ManualBundleRebuild` and handle it differently when `bundle_images` is provided or not, I can investigate on simplifying the workflow later, but we'll need to do db schema migration and alter the records in db if we have a new event now and then remove it later.

apaplaus replied:
Ok, I will try to make it like that. I will ping you again when it will be ready.

qixiang commented on freshmaker/events.py:
```diff
@@ -462,13 +462,30 @@ def __init__(self, msg_id, advisory, dry_run=False):
+        event = cls(msg_id, **kwargs)
+        event.advisory = advisory
+        event.freshmaker_event_id = freshmaker_event_id
+        return event
```
Like the case of manual rebuild of rpm advisory, `event.container_images` should also be supported, in some cases, we just need to rebuild some of the impacted bundle images.

apaplaus commented:
@qixiang Well I've made some changes, so can you check pls?

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -127,6 +129,14 @@ def _handle_auto_rebuild(self, db_event):
 
+        # Ignore builds from dependent events that were rebuilt recently
+        for orig_nvr, rebuilt_nvr in nvrs_mapping.copy().items():
+            done_build = db_event.get_artifact_build_from_event_dependencies(rebuilt_nvr)
```
This is a little confusing, the `nvrs_mapping` is a mapping of images in advisory and their original image NVRs, they're operator or operand images, and the images to skip (`event.container_images`) is the bundle images in the result of images to be rebuilt in this handler, could you correct me if I'm wrong? I believe the skipping feature is only possible to be done after the rebuilt list is figured out and before recording them in db.

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -655,33 +665,43 @@ def _create_original_to_rebuilt_nvrs_map(self):
-                            block_build_nvr['version'] == parsed_build_nvr['version']:
-                        nvrs_mapping[block_build] = build_nvr
+        # If specific container images where requested to rebuild, process only them
+        if self.event.container_images:
```
Same as above, the `event.container_images` are actually bundle images, and I believe here we're adding these bundle images to the nvr mapping of operator/operand images.

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -237,9 +239,24 @@ def _handle_auto_rebuild(self, db_event):
+            bundle_nvr = bundles[0]['brew']['build']
+
+            # If specific container images where requested to rebuild, process only them
+            if bundle_nvr in self.event.container_images:
```
it should be `not in`, right?

apaplaus replied:
Yes)

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -237,9 +239,24 @@ def _handle_auto_rebuild(self, db_event):
+
+            # If specific container images where requested to rebuild, process only them
+            if bundle_nvr in self.event.container_images:
+                self.log_debug("Ignoring '%s' bundle, because it's not in 'container_images'",
```
```suggestion
                self.log_debug("Ignoring '%s', because it's not in requested rebuilds (container_images in request)",
```

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -237,9 +239,24 @@ def _handle_auto_rebuild(self, db_event):
+            done_build = db_event.get_artifact_build_from_event_dependencies(
+                bundle_nvr)
+            if done_build:
+                self.log_debug("Ignoring '%s' bundle, because it was already rebuilt",
```
```suggestion
                self.log_debug("Ignoring '%s', because it was already rebuilt in dependent event",
```

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -677,8 +694,8 @@ def _create_original_to_rebuilt_nvrs_map(self):
                     block_build_nvr = parse_nvr(block_build)
-                    if block_build_nvr['name'] == parsed_build_nvr['name'] and \
-                            block_build_nvr['version'] == parsed_build_nvr['version']:
+                    if block_build_nvr['name'] == parsed_build_nvr['name']\
+                            and block_build_nvr['version'] == parsed_build_nvr['version']:
```
Per PEP8, wrapping in parentheses should be used in preference to use a backslash for line continuation.

```suggestion
                    if (
                        block_build_nvr["name"] == parsed_build_nvr["name"]
                        and block_build_nvr["version"] == parsed_build_nvr["version"]
                    ):
```

qixiang commented on freshmaker/models.py:
```diff
@@ -66,7 +66,7 @@
     FreshmakerAsyncManualBuildEvent: 14,
     BotasErrataShippedEvent: 15,
-    ManualBundleRebuild: 16,
+    ManualBundleRebuild: 16
```
This change is unnecessary, and per PEP8, a trailing comma is preferred.

qixiang commented on freshmaker/parsers/internal/manual_rebuild.py:
```diff
@@ -21,7 +21,8 @@
 import time
 from freshmaker.parsers import BaseParser
-from freshmaker.events import ManualRebuildWithAdvisoryEvent, ManualBundleRebuild
+from freshmaker.events import ManualRebuildWithAdvisoryEvent, \
+    ManualBundleRebuild
```
This is a unnecessary change.

qixiang commented on freshmaker/handlers/botas/botas_shipped_advisory.py:
```diff
@@ -237,9 +240,25 @@ def _handle_auto_rebuild(self, db_event):
+
+            # If specific container images where requested to rebuild, process only them
+            if (isinstance(self.event, ManualBundleRebuild)
+                    and bundle_nvr not in self.event.container_images):     # noqa: W503
```
You need to check if `container_images` is empty or not, otherwise all bundles will be ignored in case of `container_images` is not specified.

apaplaus replied:
fixed

qixiang commented:
LGTM
